services/src/main/java/org/keycloak/authentication/requiredactions/VerifyUserProfile.java

```diff
@@ -86,60 +85,24 @@ public void requiredActionChallenge(RequiredActionContext context) {
                 parameters = context.getHttpRequest().getDecodedFormParameters();
             }
 
-            context.challenge(createResponse(context, profile, parameters, errors));
-        }
-    }
-
-    @Override
-    public void processAction(RequiredActionContext context) {
-        EventBuilder event = context.getEvent();
-        event.event(EventType.VERIFY_PROFILE);
-        MultivaluedMap<String, String> formData = context.getHttpRequest().getDecodedFormParameters();
-
-        if (!context.getRealm().isEditUsernameAllowed()) {
-            formData.putSingle(UserModel.USERNAME, context.getUser().getUsername());
-        }
-
-        UserProfileProvider provider = context.getSession().getProvider(UserProfileProvider.class);
-        UserProfile profile = provider.create(UserProfileContext.UPDATE_PROFILE, formData, context.getUser());
-
-        try {
-            profile.update();
-            context.success();
-        } catch (ValidationException ve) {
-            List<FormMessage> errors = Validation.getFormErrorsFromValidation(ve.getErrors());
-            context.challenge(createResponse(context, profile, formData, errors));
+            context.challenge(createResponse(context, parameters, errors));
+            
+            EventBuilder event = context.getEvent().clone();
+            event.event(EventType.VERIFY_PROFILE);
+            event.detail("fields_to_update", collectFields(errors));
+            event.success();
         }
     }
 
-
-    @Override
-    public void close() {
-
+    private String collectFields(List<FormMessage> errors) {
+        return errors.stream().map(e -> e.getField()).distinct().collect(Collectors.joining(","));
```

Comment:
Could you use method reference here instead?

Reply:
Sure, too new concept for me, I'm still forgeting about this possibility ;-)

Reply:
updated